### PR TITLE
Upgrades to CoordinateSystem and graphing.

### PR DESCRIPTION
**Pull request by hydrobeam**
## Changelog/Overview

Breaking changes were introduced to :class:`Axes`, :class:`ThreeDAxes`, :class:`NumberPlane` and :class:`NumberLine`
All the above now use lists to construct their ranges as opposed to explicitly defining these values --> `x_range`  :  `[x_min, x_max, x_step]`. There were also many upgrades to these classes which improves their functionality and ease of use.

Furthermore, `~.GraphScene` has been deprecated and we encourage use of the above functions instead. It can still be used because :class:`NumberLineOld` was introduced to accommodate these changes.

## Motivation

The current handling of graphing logic with `coordinate_systems.py` was outdated and not very intuitive. This PR ports over the changes made upstream to `coordinate_systems.py`, `numberline.py` and `functions.py` which should make the process of graphing less convoluted. 

### `coordinate_systems.py` 
- Rework `CoordinateSystem` to accommodate `x_range` and `y_range` and their corresponding length values.
   - Ranges represent ``[min, max, step]``
   - `get_graph` updated to support the range.
   - `get_line_from_axis_to_point` introduced.
- Rework `Axes`
  - Uses `x_range` and `y_range` to create the axes.
  - default `x_length` and `y_length` changed so that the axes are each two units away from the edges of the screen (see below)
image
   - `create_axis` automatically shifts the axes based on the zero point (if it does not exist, shifts to the max/min)
   - `stroke_width` reduced to `2` resulting in thinner axes.
   - `ThreeDAxes` compatibility with Axes.
- `NumberPlane` compatiblity with Axes
   - `NumberPlane`, is now enclosed by lines(box shaped) and is smaller.
   - The plane adjusts  its size based on the `x_range` and `y_range` value so that the unit size is always one (represents coordinates).
   - New
image
   - Old
image

- `ComplexPlane` support 
 
### `number_line.py`
- `x_min`, `x_max`, `add_start`, `add_end`, `tick_frequency`, `leftmost_tick`, `number_at_center`, `numbers_to_show` (might re-introduce `numbers_to_show`) removed
- `numbers_to_exclude` replaces `exclude_zero_from_default_numbers`, `length` replaces `width`
- `stroke_width` and `x_range` added.
-  Redundant methods removed, some methods renamed to be clearer, other methods reworked/updated. (such as `add_numbers`)
- `UnitInterval` updated to reflect changes.

### `functions.py`
- Accommodate using ranges.

### `graph_scene.py`
- Deprecated.
- OldNumberLine created in `number_line.py` to accommodate these changes.

Code to recreate default GraphScene axes:
```python
class AxesShowcase(Scene):
    def construct(self):
        graph = Axes(
            x_range=[-1,10,1],
            y_range=[-1,10,1],
            x_length=9,
            y_length=6,
            axis_config={"include_tip":False}
        )
        labels = graph.get_axis_labels()
        self.add(graph, labels)
```
image

## Testing Status

I modified some tests to accommodate the changes. This included removing the usage of `GraphScene` for some tests and using `Axes` for some tests, but generally they should all produce the same output, or at least similar. The test that was entirely reworked was the NumberPlane test since it involved splitting `x_min` and `x_max` between the `axis_config`s and attributes, which is not possible with the new range system.

## Further Comments

A bit of the code that was added is currently redundant (see: `font_size` in `NumberLine` for example) because it relies on updated functionality from other files. I'm not sure what the stance is on updating those, so I left it alone. 

It is worth noting that trying to recreate some of the code in the documentation and tests using solely `Axes` was at times janky, so I might also port over some tools used by `GraphScene` into `Axes`. Although, the code was much cleaner and simpler for most tasks.
 
I will be adding the documentation changes in the next PR.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have chosen a descriptive PR title (see top of PR template for examples)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The PR title is descriptive enough

### CODE REVIEW

**Comment by GameDungeon** [comment]
Could you re-write the test you are now failing? These are some nice improvements and I'm quite happy you have made this.

File: manim/mobject/coordinate_systems.py
Comment (GameDungeon) [inline manim/mobject/coordinate_systems.py:13]: Nitpicky, but could you make the imports in line pelase.
```suggestion
from ..mobject.geometry import Arrow, DashedLine, Line
```

**Review by GameDungeon — APPROVED** [review]
LGTM

Needs typing but that should be added in a later pull. Could this be merged quickly, as it is extremely useful? I can't overstate how happy I am with the changes added here.

File: manim/mobject/coordinate_systems.py
Comment (kolibril13) [inline manim/mobject/coordinate_systems.py:123]: is this not the same like `get_vertical_line_to_graph`?
In case that not, maybe you can add an example here:
https://manimce--1161.org.readthedocs.build/en/1161/examples.html#sinandcosfunctionplot
so one can see in action what the difference between these two is

**Comment by kolibril13** [comment]
> The next step is to shift over `GraphScene`'s useful methods over to `CoordinateSystem` and deprecate `GraphScene`.

Great idea.
Thanks for this amazing addition.
One thing I just realized when looking over the old and the new docs is that the size of the axis decreased:
image
Is that intentional?

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:123]: I compared the two and I think they're a bit different: `axis_to_point` takes in a point anywhere on the graph, whereas `line_to_graph` is specifically linked to a predefined function. They're definitely at least similar. That example is already a little busy, so adding another line doesn't seem like the best idea though.

**Comment by hydrobeam** [comment]
For some reason I can't reply to the previous comment 😅. But yes the `stroke_width` change is intentional. I noticed that the upstream version used thinner axes and I was a fine of the style; but it's very easily reversible. I'm open to hearing opinions about it!

File: manim/mobject/coordinate_systems.py
Comment (kolibril13) [reply manim/mobject/coordinate_systems.py:123]: Okay, nice, that makes sense.
Then we can e.g. add it here:
https://manimce--1161.org.readthedocs.build/en/1161/examples.html#graphareaplot

**Comment by kolibril13** [comment]
I have not too strong feelings about the `stroke_width`, we can keep it like it is.
Is it possible to fine-tune the `stroke_width` for the axes with a user facing parameter?

**Comment by hydrobeam** [comment]
Yep, there's a `stroke_width` parameter included in `NumberLine` that defaults to 2.0.

**Comment by kolibril13** [comment]
I suggest that we make then one of the examples in the example gallery with thicker stroke width and choose the thin stroke width as default.

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:123]: I just realized that it actually won't work because this method is in `CoordinateSystem` and `GraphScene` does not rely on `CoordinateSystem` at all. The documentation changes should probably be saved when I port the examples to `Axes`.

**Comment by GameDungeon** [comment]
Here is an idea: Make X and Y Length a ratio. the current 9 & 6 being [3, 2]. It simplifies the args, might want others opinions on this as it might convolute things.

**Review by behackl — COMMENTED** [review]
Thank you for your efforts! I think this is already a pretty good direction, however there are some issues I'd like to have discussed before moving this further. I've left a bunch of suggestions and comments that are up for discussion.

File: manim/scene/graph_scene.py
Comment (GameDungeon) [reply manim/scene/graph_scene.py:75]: Honestly just revert the changes and deprecate it.

**Comment by GameDungeon** [comment]
If everyone agrees with me, revert the changes to GraphScene and instead just deprecate it.

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:195]: Yup, it's set by default in `NumberLine` so it was redundant in this `axis_config`

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:379]: Using `super().__init__` would cass the number plane to inherit the axes currently reduced length and not fill up the entire screen. That's why i went with the approach of re-calling it and using the entire screen for `NumberPlane`

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:563]: Not sure, I tried it with and without and it just seems to cut off one of the edge numbers:

With:
image

Without: 
image

I guess it's to keep the sides balanced in terms of numbers but poorly executed? Or maybe just a relic.

**Comment by hydrobeam** [comment]
I'll take your recommendations and create a `OldNumberLine` class with a deprecation warning. I'll revert the changes to `GraphScene` and make it so that it automatically calls this class.

**Comment by GameDungeon** [comment]
Could this be updated to the latest version?

**Comment by hydrobeam** [comment]
I've carried out the changes. `GraphScene` and its related tests/docs have been reverted to their original state and deprecated. To support `GraphScene`, I created a new class called `OldNumberLine` in `number_line.py` which uses the old system. Unfortunately, there are some conflicts and I'm not sure how to resolve them.

**Comment by hydrobeam** [comment]
After some effort, I patched up the merge conflicts. It should be good to go!

**Comment by GameDungeon** [comment]
Thanks so much :) What do you think @behackl now? It LGTM to me.

File: manim/mobject/geometry.py
Comment (behackl) [inline manim/mobject/geometry.py:843]: Added punctuation. Also, a test would be nice -- either as a doctest (like in the suggestion), or as a separate (non-graphical) unit test in `tests`.

```suggestion
        """Return the projection of a point onto the line.

        Examples
        --------

        ::

            >>> import numpy as np
            >>> line = Line(LEFT, RIGHT)
            >>> line.get_projection(np.array([0, 1, 0]))
            array([0, 0, 0])

```

File: manim/mobject/number_line.py
Comment (behackl) [reply manim/mobject/number_line.py:23]: The way you've changed things right now still introduces a major breaking change to `NumberPlane`, and people will have to adapt their code to use `NumberPlaneOld` on their own -- it's the opposite direction of what I suggested above.

I won't block the PR over it, but I would certainly like to hear the opinions of other community devs before proceeding with this solution.

**Review by behackl — COMMENTED** [review]
Two more comments.

Due to the size of this PR it's unfortunately a bit difficult to review properly, but I more or less agree with the current status of the refactor. (It's definitely an improvement over the current situation.)

Before moving this forward, I'd like to get some additional opinions by other community devs as well. Thank you for your efforts!

**Comment by GameDungeon** [comment]
Going off what behackl said change the old numberplane back to numberplane. Deprecate it. And then add numberplanenew.

File: manim/mobject/number_line.py
Comment (hydrobeam) [reply manim/mobject/number_line.py:23]: Even though both `GraphScene` and `NumberPlane` are widely used, switching code over from `GraphScene` to `Axes` requires significant effort, so I felt that creating a separate `OldNumberLine` was reasonable. The changes to `NumberPlane` *are* breaking, but I think that it's comparably very easy to adjust the code to get it back up and running. That's why I'm a little hesitant about adding a separate `NumberPlaneOld` class.

**Comment by Darylgolden** [comment]
Since we're deprecating `GraphScene`, we should definitely be replacing the examples in the docs that use it. It doesn't have to be done in this PR, but it should be done before the next release (it won't be nice to have 4 examples in our Example gallery made with a deprecated class). Thank you very much for the contribution, I'll be looking forward to using it!

**Comment by hydrobeam** [comment]
I can take care of recreating existing scenes in the docs with `Axes` instead in a new PR, it'd just be impossible to review properly if I added them here😅. Plus, it'd be a good opportunity to try and see what Axes is missing and could use some work copying over from `GraphScene`.

**Comment by GameDungeon** [comment]
Is this good now then?

File: manim/mobject/coordinate_systems.py
Comment (abhi3019) [inline manim/mobject/coordinate_systems.py:177]: In my opinion these lengths are really unnecessary because :
1. they seem very random
2. they can be passed via `x/y_axis_config = {"length": <length>}` or `axis_config = {"length": <length>}`
3. `unit_size` of the `x_axis` and `y_axis` is no longer respected (became redundant) because both `length` and `unit_size` shouldn't be passed into `NumberLine`. Try this code for example:
```py
plane = NumberPlane(
            x_range=[-2, 4],
            y_range=[-2, 3],
            axis_config={"include_numbers": True, "unit_size": 0.5},
        )
# print([axis.unit_size for axis in plane.axes]) --> [2.3703703703703702, 1.6]
self.add(plane)
```
No matter what value you pass for `unit_size` via `axis_config` or `x/y_axis_config`, you'll always get the same result.
For the above reasons, I'd suggest to remove the length parameters

File: manim/mobject/coordinate_systems.py
Comment (abhi3019) [inline manim/mobject/coordinate_systems.py:137]: Any specific reason you chose to remove the `center_point` parameter? It's very useful imo

File: manim/mobject/coordinate_systems.py
Comment (abhi3019) [inline manim/mobject/coordinate_systems.py:221]: Not really sure what's happening here, but to me it looks like this logic should be implemented in `NumberLine` itself

File: manim/mobject/geometry.py
Comment (abhi3019) [inline manim/mobject/geometry.py:841]: This is a generally useful function and shouldn't be tied to `Line` imo. Can we move this into `manim\utils\space_ops.py`? It would look like [this](https://github.com/abhi3019/manim-projects/blob/2e831d76ecaa06b37beda00ea0d4c767be24f372/custom_functions/manim.py#L5-L12)

File: manim/mobject/number_line.py
Comment (abhi3019) [inline manim/mobject/number_line.py:90]: Not a big issue, but we can make use of `Line.set_length` instead. It does the same thing and sounds more intuitive imo
```suggestion
            self.set_length(self.length)
```

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:177]: 1. The lengths are specified so that the axes don't take up the entire screen to make them look more visually appealing, but I'm not sure what you mean by random.
2. I think the length of the axes is a very important property to be able to easily modify. Working with axis configs is generally more tedious as opposed to an explicit parameter. There *is* a little redundancy, but I find the current implementation more intuitive.
3. I hadn't noticed this redundancy, but this makes me think that `unit_size` should be removed, not `x_length` and `y_length`.

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:137]: The changes under `create_axis` make it so that the "`center_point`" is determined dynamically, making it so that defining `center_point` was a bit redundant. Plus, the axes can also be shifted after creation if necessary.


Without `center_point`: 
image

with the previous default ```center_point = ORIGIN```:
image

File: manim/mobject/coordinate_systems.py
Comment (abhi3019) [reply manim/mobject/coordinate_systems.py:177]: >I'm not sure what you mean by random

I just meant that the `-2` part felt random, as in it could've been any other number. But we have to use some default length anyway, so that's fine



> 3\. this makes me think that `unit_size` should be removed, not `x_length` and `y_length`

None of those should be compromised in favor of the other imo. Both are extremely useful and removing one of them doesn't look good to me. Maybe instead of `x_length` and `y_length` we can introduce the parameters `x_unit` and `y_unit`? It might not seem to make much difference, but it addresses the following points:

4. The default `NumberPlane` has its `x_axis` stretched out a little
```py
plane = NumberPlane()
print(plane.x_axis.unit_size, plane.y_axis.unit_size) 
>>> 1.0158730158730158 1.0
```
As far as I've seen (and used myself), people often use the default `NumberPlane` as a reference to manim's default coordinate system. With the current implementation, users are forced to do `NumberPlane([-config.frame_x_radius, config.frame_x_radius])` to represent the default coord-system. This also forces the use of `plane.c2p(<coords>)` instead of just `coords`. Although it is a good practice to use `plane.c2p`, it seems a bit too much work while working with default coord-system.

5. Setting `unit_size` becomes difficult
If the user wants to set a specific unit size (extremely useful with `LinearTransformation` demos), its tedious with current implementation
```py
x_min, x_max = -4, 4
x_unit = 2
plane = NumberPlane([x_min, x_max], x_length=(x_max - x_min) * x_unit)  
```

6. Visual appeal (also notice something is wrong with the `NumberLine`'s axis ticks:
```py
# master
# plane = NumberPlane(x_min=-5, x_max=6, y_min=-3, y_max=2)
# line = NumberLine(x_min=-config.frame_x_radius, x_max=config.frame_x_radius)
                
# pr branch
plane = NumberPlane([-5, 6], [-3, 2])
line = NumberLine([-config.frame_x_radius, config.frame_x_radius])

line.shift(-line.n2p(0)).shift(DOWN)
line_origin = Dot(line.n2p(0), color=YELLOW)

self.add(plane.set_opacity(0.5), line, line_origin)
```
on master:
CoordinateSystemsTest_ManimCE_v0 5 0
on pr branch:
CoordinateSystemsTest_ManimCE_v0 5 0
It doesn't really look good to me that the plane is always stretched out (and with the origin shifted) and you'll have to do extra work setting the `length` as `(max-min)*unit` to set some `unit_size`

Some points might just look like my personal preferences, I'd like to hear others opinions as well :)

File: manim/mobject/coordinate_systems.py
Comment (abhi3019) [reply manim/mobject/coordinate_systems.py:137]: > The changes under create_axis make it so that the "center_point" is determined dynamically

I see, that's nice

> Plus, the axes can also be shifted after creation if necessary

It has to be done like `axes.shift(target_origin - axes.c2p(0, 0))`. Again, this looks like a bit too much work. I'm not against removing the `center_point` parameter, but if the preference is to drop it, let's at least have a method to shift the axes's origin to some other point. Something like
```py
def move_origin_to(self, new_origin, *other_mobs):
    for mob in [self, *other_mobs]:
        mob.shift(new_origin - self.c2p(0, 0))
```
This could probably be done in a later PR, just leaving my thoughts.

Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:137]: I think this would be better considered in a separate PR.

**Comment by hydrobeam** [comment]
@abhi3019 pointed out that the default `NumberPlane` was stretched so that it did not properly map to the default coordinates. I edited the length values of `NumberPlane` so that it now does this. This does leave the basic `NumberPlane` with some extra space on the sides, but I think it's a worthwhile trade-off. I did consider reverting to the previous method of creating `NumberPlane` background lines (without the box look), but the current implementation looks significantly better for resized planes.

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py]: I looked into the issue with the misplaced ticks on the number line and it arises due to the the ticks being drawn starting from the left and continuing onwards. This makes it seem like the dot is misplaced based on tick placement but it's still at 0.

image

As for the odd looking default axis. It would be possible to make it so that x_length and y_length are by default ```x_max-x_min``` so that `unit_size` is always one initially. However, it should be rather intuitive for a user that if `x_range =[-5, 5, 1]` that an x_length of 10 would result in a unit_size of one, or `x_length = 5` would make the unit size two.

**Comment by hydrobeam** [comment]
I've made it so that `NumberPlane` `x_length` and `y_length` scale automatically when not specified to keep the unit size at 1. Plus, the default values for `NumberPlane` are now smaller to accomodate the box look: 

image

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:221]: I did some light testing and in the current iteration it belongs better in `create_axis`. Shifting this functionality to `NumberLine` would take some work restructuring for little benefit since this is specifcally tuned for `Axes`.

File: manim/mobject/number_line.py
Comment (behackl) [reply manim/mobject/number_line.py:23]: It's fine, we can stick with introducing a breaking change to `NumberPlane`. In this case please:

1. update the PR description: within the 
   ```
   <!--changelog-start-->
   <!--changelog-end-->
   ```
   block, add a short migration guide from the old to the new usage of `NumberPlane`, and remark that the old mobject is still available via `NumberPlaneOld`.
2. Maybe also add a `logger.info` making users aware of the fact that this is the new interface and that the old class is available as `NumberPlaneOld`.

File: manim/mobject/number_line.py
Comment (behackl) [inline manim/mobject/number_line.py:24]: I'd really like to see a basic docstring for the `NumberLine` class; it doesn't have to include examples -- but a `Parameters`-section where each of the available parameters is briefly explained would be great.

**Review by behackl — CHANGES_REQUESTED** [review]
I think this PR has been open for long enough, I'd like to get this merged and move on.

I've requested two more changes, but am in favor of merging this afterwards.

File: manim/mobject/number_line.py
Comment (hydrobeam) [reply manim/mobject/number_line.py:23]: There actually isn't a `NumberPlaneOld` at the moment, just a `NumberLineOld` to work with `GraphScene`. Would you like me to add a `NumberPlaneOld`?

Comment (behackl) [reply manim/mobject/number_line.py:23]: No, I meant NumberLine / NumberLineOld.

**Comment by GameDungeon** [comment]
I agree. I LGTM with those changes as well. Thanks a lot Hydrobeam for all the work you put in.

**Comment by hydrobeam** [comment]
I've updated the changelog in the PR description and added some documentation for the classes affected by these changes. Let me know if there's anything else that needs to be added.

File: manim/mobject/number_line.py
Comment (behackl) [inline manim/mobject/number_line.py:67]:
```suggestion
        Additional arguments to be passed to :class:`~.Line`.
```

File: manim/mobject/number_line.py
Comment (behackl) [inline manim/mobject/number_line.py:71]:
```suggestion
    NumberLine
        The constructed number line.
```

File: manim/mobject/number_line.py
Comment (behackl) [inline manim/mobject/number_line.py:302]:
```suggestion
    """The old version of :class:`NumberLine`. 
    
    Exists for compatibility with :class:`~.GraphScene`.
    """
```

**Review by behackl — CHANGES_REQUESTED** [review]
Thanks for adding documentation! I've looked through it and added a few comments/suggestions, please have a look.

And as a further remark: as soon as we have #1404, the deprecation should be changed to use the new decorators. This can be done in a later PR though.

File: manim/mobject/coordinate_systems.py
Comment (hydrobeam) [reply manim/mobject/coordinate_systems.py:233]: It means 'where 0 is located on the line". I guess it would be clearer to leave it at `adjusts its position`.

**Review by behackl — APPROVED** [review]

**Comment by behackl** [comment]
Lets merge this. In case something seriously breaks, we can always do a bugfix release.

Thank you for your efforts -- and your patience. This took a bit longer than I would have liked.

**Comment by kolibril13** [comment]
Nice!
I just had a look on the docs, maybe we can update it here as well:
https://docs.manim.community/en/latest/reference/manim.scene.graph_scene.html?highlight=graph_scene
and here:
https://docs.manim.community/en/latest/examples.html#graphareaplot
there seems to be a broken link
image

**Comment by behackl** [comment]
> maybe we can update it

What exactly would you like to have updated? The examples?

**Comment by kolibril13** [comment]
````
     y_min=0,
     y_max=100,
     y_axis_config={"tick_frequency": 10},
```
this part

**Pull request merged**
